File: src/PicLens.Core/Domain/SettingsRules.cs
using PicLens.Core.Models;

namespace PicLens.Core.Domain;

public static class SettingsRules
{
    public const int DefaultThumbnailSize = 200;
    public const int MinThumbnailSize = 120;
    public const int MaxThumbnailSize = 240;
    public const int ThumbnailSizeStep = 20;

    public static AppSettings NormalizeSettings(AppSettings settings)
    {
        var thumbnailSize = settings.ThumbnailSize == 0
            ? DefaultThumbnailSize
            : NormalizeThumbnailSize(settings.ThumbnailSize);

        return settings with
        {
            ThumbnailSize = thumbnailSize
        };
    }

    public static int NormalizeThumbnailSize(double thumbnailSize)
    {
        if (double.IsNaN(thumbnailSize) || double.IsInfinity(thumbnailSize))
        {
            return DefaultThumbnailSize;
        }

        var stepped = (int)(Math.Round(thumbnailSize / ThumbnailSizeStep, MidpointRounding.AwayFromZero) * ThumbnailSizeStep);
        return Math.Clamp(stepped, MinThumbnailSize, MaxThumbnailSize);
    }

    public static AppSettings MergeSettingsPatch(AppSettings current, AppSettingsPatch patch)
    {
        var normalizedCurrent = NormalizeSettings(current);

        return normalizedCurrent with
        {
            LastFolderPath = patch.HasLastFolderPath ? patch.LastFolderPath : normalizedCurrent.LastFolderPath,
            Sort = patch.Sort ?? normalizedCurrent.Sort,
            IncludeSubfolders = patch.IncludeSubfolders ?? normalizedCurrent.IncludeSubfolders,
            ThumbnailSize = patch.ThumbnailSize.HasValue
                ? NormalizeThumbnailSize(patch.ThumbnailSize.Value)
                : normalizedCurrent.ThumbnailSize
        };
    }
}

File: src/core/src/settings_rules.cpp
#include <piclens/core/settings_rules.h>

#include <algorithm>
#include <cmath>

namespace piclens::core::settings_rules {

AppSettings normalizeSettings(const AppSettings &settings)
{
    AppSettings normalized = settings;
    normalized.thumbnailSize = settings.thumbnailSize == 0
        ? DefaultThumbnailSize
        : normalizeThumbnailSize(settings.thumbnailSize);
    return normalized;
}

int normalizeThumbnailSize(double thumbnailSize)
{
    if (!std::isfinite(thumbnailSize)) {
        return DefaultThumbnailSize;
    }

    const int stepped = static_cast<int>(std::round(thumbnailSize / ThumbnailSizeStep))
        * ThumbnailSizeStep;
    return std::clamp(stepped, MinThumbnailSize, MaxThumbnailSize);
}

AppSettings mergeSettingsPatch(const AppSettings &current, const AppSettingsPatch &patch)
{
    AppSettings merged = normalizeSettings(current);
    if (patch.hasLastFolderPath) {
        merged.lastFolderPath = patch.lastFolderPath;
    }
    if (patch.sort.has_value()) {
        merged.sort = *patch.sort;
    }
    if (patch.includeSubfolders.has_value()) {
        merged.includeSubfolders = *patch.includeSubfolders;
    }
    if (patch.thumbnailSize.has_value()) {
        merged.thumbnailSize = normalizeThumbnailSize(*patch.thumbnailSize);
    }
    return merged;
}

} // namespace piclens::core::settings_rules
